### PR TITLE
runtime.onInstalled use in event pages clarification

### LINKED ISSUE #17212: Example code works via web-ext run but not after packaging into an XPI
### MDN URL

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Background_scripts

### What specific section or headline is this issue about?

Add listeners

### What information was incorrect, unhelpful, or incomplete?

The example code on this page shows a context menu being added inside an onInstalled listener. This MDN page came up when searching Google on how to add a context menu to an extension. I based my code on the example on this page. Initially it worked perfectly while I was testing the extension via web-ext run but once I packaged it into an XPI the context menu was no longer created.

### What did you expect to see?

Code example that adds a context menu all the time not just after installation.

### Do you have any supporting links, references, or citations?

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onInstalled

### Do you have anything more you want to share?

The onInstalled MDN page says: "Fired when the extension is first installed, when the extension is updated to a new version, and when the browser is updated to a new version". However this page says in regards to onInstalled: "For extensions with event pages, this is where stateful APIs, such as a context menu created using browser.menus.create, should be used". These two seem to contradict each other, the first implying it only runs once after install, and the second implying it runs every time the event page is loaded. Assuming the first explanation is correct then this example is obviously wrong, but either way this contradiction could do with some more clarity in the text.

FWIW after installing my XPI I never once saw my context menu, even directly after I installed the XPI, don't know why. Once I created the context menu directly in the background script (not in a onInstalled handler) everything worked fine.

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/mozilla/add-ons/webextensions/background_scripts`
* MDN URL: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Background_scripts
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/mozilla/add-ons/webextensions/background_scripts/index.md
* Last commit: https://github.com/mdn/content/commit/3320c01d3f01107d9e4fc2f98160183fb27e913e
* Document last modified: 2022-05-15T01:48:39.000Z

</details>

### PR DESCRIPTION
**Pull request by rebloor**
### Description

Add an explanation as to why stateful API should be included in the onInstalled for event pages. Fixed #17212

### CODE REVIEW

**Review by Rob--W — CHANGES_REQUESTED** [review]
I think that more details are needed before #17212 can be marked as resolved.

The Extension Workshop has a lot of detail on the persistence of menu items across restarts. The short version is that in Firefox, menu items are not persisted in persistent background pages (a persistent background page starts up once and doesn't terminate until the browser or extension unloads, so a top-level `create` call will always create the menu), and persisted in non-persistent background pages (aka event page; to avoid repeated attempts to create the menu item, it should be registered only once).

Can you clarify the documentation to reduce this confusion? E.g. by adding a NOTE and more details (maybe even a link to Extension workshop?) in the menus.create article on MDN.

File: files/en-us/mozilla/add-ons/webextensions/api/menus/create/index.md [hunk 1]
```diff
@@ -7,11 +7,20 @@ browser-compat: webextensions.api.menus.create
...
+For compatibility with other browsers, Firefox makes this method available in the `contextMenus` namespace and `menus` namespace. However, it's not possible to create tools menu items (`contexts: ["tools_menu"]`) using the `contextMenus` namespace.
+
+> **Creating menus in persistent and non-persistent extensions**
+>
+> How you create menu items depends on whether your extension is persistent or non-persistent:
+>
+> - in extensions using persistent background pages, menus are not persisted. A persistent background page starts once and doesn't terminate until the browser or extension unloads. So, you use a top-level call to create the menu.
+> - in extensions using non-persistent background pages (an event page), menus persist. You call `menus.create` from within a {{WebExtAPIRef("runtime.onInstalled")}} listener. This avoids repeated attempts to create the menu item when the pages restart, which would occur with a top-level call.
```
Comment (Rob--W) [inline files/en-us/mozilla/add-ons/webextensions/api/menus/create/index.md:21]:
```suggestion
> - in extensions using non-persistent background pages (an event page), menus persist across browser and extension restarts. You call `menus.create` (with a menu-specific ID) from within a {{WebExtAPIRef("runtime.onInstalled")}} listener. This avoids repeated attempts to create the menu item when the pages restart, which would occur with a top-level call.
> - In Chrome, menu items from persistent background pages are also persisted.
> - In Firefox, menu items from persistent background pages are never persisted. Call `menus.create` unconditionally from the top-level to make sure that the menu items are registered.
```

I swapped the bullet points and emphasized the difference between Chrome and Firefox.

File: files/en-us/mozilla/add-ons/webextensions/api/menus/index.md [hunk 2]
```diff
@@ -21,6 +21,8 @@ Except for [`menus.getTargetElement()`](/en-US/docs/Mozilla/Add-ons/WebExtension
 
 To create a menu item call the {{WebExtAPIRef("menus.create()")}} method. You pass this method an object containing options for the item, including the item ID, item type, and the contexts in which it should be shown.
 
+In an extension using non-persistent [background pages](/en-US/docs/Mozilla/Add-ons/WebExtensions/Background_scripts) (Event pages), you call `menus.create` from within a {{WebExtAPIRef("runtime.onInstalled")}} listener. In an extension using persistent background pages, you make a top-level call. See {{WebExtAPIRef("menus.create()")}} for more information.
```
Comment (Rob--W) [inline files/en-us/mozilla/add-ons/webextensions/api/menus/index.md:24]: The behavior is Firefox/Chrome-dependent.

**Review by Rob--W — CHANGES_REQUESTED** [review]

**Review by Rob--W — APPROVED** [review]

**Comment by rebloor** [comment]
Thanks @Rob--W

**Pull request merged**